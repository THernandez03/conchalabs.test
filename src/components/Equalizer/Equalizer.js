import { css } from '@emotion/react';
import styled from '@emotion/styled';

const cssWrapper = css`
  display: flex;
  flex-grow: 1;
  align-items: center;
  justify-content: center;
`;

const EqualizerInput = styled(({ className }) => (
  <input
    className={className}
    type="range"
    orient="vertical"
    min="-40"
    max="40"
  />
))`
  transform: rotate(-90deg);
`;

export const Equalizer = () => (
  <div css={cssWrapper}>
    <EqualizerInput />
    <EqualizerInput />
    <EqualizerInput />
  </div>
);
