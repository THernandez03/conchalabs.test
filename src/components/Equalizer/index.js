export * from './Equalizer';
export * from './EqualizerSlider';
