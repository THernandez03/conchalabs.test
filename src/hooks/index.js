export * from './useBiquadFilter';
export * from './useMediaSource';
