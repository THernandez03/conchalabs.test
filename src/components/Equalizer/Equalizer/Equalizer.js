import { Children, cloneElement, useEffect, useState } from 'react';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';

import { useAudioDependencies } from '../../../hooks';

const InnerWrapper = styled.div`
  display: flex;
  flex-grow: 1;
  justify-content: space-evenly;
  max-width: 768px;
  margin: 0 auto;
`;

export const Equalizer = styled(({ className, children }) => {
  const [sliders, setSliders] = useState();
  const { audioContext, mediaSource } = useAudioDependencies();

  useEffect(() => {
    if (!audioContext || !mediaSource || !children) return;

    const cloned = Children.map(children, (child) =>
      cloneElement(child, {
        filter: audioContext.createBiquadFilter(),
      }),
    );

    mediaSource.disconnect();
    const filter = cloned.reduce(
      (acc, child) => acc.connect(child.props.filter),
      mediaSource,
    );
    filter.connect(audioContext.destination);

    setSliders(cloned);
  }, [audioContext, mediaSource, children]);

  return (
    <div className={className}>
      <InnerWrapper>{sliders}</InnerWrapper>
    </div>
  );
})`
  display: flex;
  flex-grow: 1;
  align-items: center;
  background: #2980b9;
  background: -webkit-linear-gradient(to right, #2c3e50, #2980b9);
  background: linear-gradient(to right, #2c3e50, #2980b9);
`;

Equalizer.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  className: PropTypes.string,
};

Equalizer.defaultProps = {};
