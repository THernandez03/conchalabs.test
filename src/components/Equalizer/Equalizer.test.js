/* eslint-disable import/unambiguous */

describe('<Equalizer/>', () => {
  it('test', () => {
    expect(true).toBe(true);
  });
});
