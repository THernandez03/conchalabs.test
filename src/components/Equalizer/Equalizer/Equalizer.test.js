import { render } from '@testing-library/react';

import { Equalizer } from './Equalizer';
import { useAudioDependencies } from '../../../hooks';

jest.mock('../../../hooks', () => ({
  useAudioDependencies: () => ({}),
}));

describe('<Equalizer/>', () => {
  it('Should render the component', () => {
    const { container } = render(<Equalizer />);
    expect(container).toMatchSnapshot();
  });
});
