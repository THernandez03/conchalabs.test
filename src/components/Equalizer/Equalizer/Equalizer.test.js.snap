// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Equalizer/> Should render the component 1`] = `
<div>
  <div
    class="css-h3x0xf-Equalizer e4ja4jf0"
  >
    <div
      class="css-1ee0tmd-InnerWrapper e4ja4jf1"
    />
  </div>
</div>
`;
